### PR TITLE
[SDK] Support default value specified in flow.flex.yaml & type validation for flex flow

### PR DESCRIPTION
**Pull request by D-W-**
# Description

Please add an informative description that covers that changes made by the pull request and link all relevant issues.

This pull request includes changes to improve the handling of signatures and flow context in the `promptflow` package. The most important changes include the introduction of a new `flow_overwrite_context` function to replace `variant_overwrite_context`, the addition of a new `test_update_signature` function in the `test_flow_entity.py` file, and the modification of the `update_signatures` function in the `signature_utils.py` file.

Improvements to flow context handling:

* [`src/promptflow-devkit/promptflow/_sdk/_orchestrator/__init__.py`](diffhunk://#diff-d6601fab1c3e78da457ecfc79f0f87797137bf11299ab47911be7f8a896ba18dR7-R18): Renamed `variant_overwrite_context` to `flow_overwrite_context` and updated all references in other files. [[1]](diffhunk://#diff-d6601fab1c3e78da457ecfc79f0f87797137bf11299ab47911be7f8a896ba18dR7-R18) [[2]](diffhunk://#diff-eb5aea25b8997068780361beb30138aaad86db060b97fe9673585ef258cd8b4bL28-R28) [[3]](diffhunk://#diff-d5ded9a013487d37d3519fcb8c5433381c64ecc0d922629e87429b645a2e9d35R38-L41) [[4]](diffhunk://#diff-bad1e9183cc8452c12df1c0932a6a10c7c8224382761a4a6702a8737981da5f9L205-R206) [[5]](diffhunk://#diff-03522f1c332af4cf8ec836d2d6a3ea2c7c84f29936dd98cf24539b83b4f47993L122-R122) [[6]](diffhunk://#diff-a6386c44a6e32dbef95fff7e5fcdeb933e5fdea154369d37d7cec10bda0eb226L528-R535) [[7]](diffhunk://#diff-a73c2e1869aeb231ae39aad23e3b85155c1b1228836624a33b558bbd59e54b18L14-R14)

* [`src/promptflow-devkit/promptflow/_sdk/_orchestrator/utils.py`](diffhunk://#diff-bad1e9183cc8452c12df1c0932a6a10c7c8224382761a4a6702a8737981da5f9R218): Updated the `flow_overwrite_context` function to handle `FlexFlow` instances and added a call to `update_signatures` for `FlexFlow` and `Flow` instances. [[1]](diffhunk://#diff-bad1e9183cc8452c12df1c0932a6a10c7c8224382761a4a6702a8737981da5f9R218) [[2]](diffhunk://#diff-bad1e9183cc8452c12df1c0932a6a10c7c8224382761a4a6702a8737981da5f9R227-R245)

* [`src/promptflow-devkit/promptflow/_sdk/operations/_local_storage_operations.py`](diffhunk://#diff-2e3d235b95e35f753f93d9ab53a208554645da4a6a45191cc7930d3b1a43283aL259-R260): Removed the call to `update_signatures` in the `dump_snapshot` function.

Improvements to signature handling:

* [`src/promptflow-azure/tests/sdk_cli_azure_test/unittests/test_flow_entity.py`](diffhunk://#diff-78df84c4c6e9e0905dd6dad83698db0e0d10113bee788bf552dfc2a1976bc8c2L312-R320): Added the `test_update_signature` function to test the `update_signatures` function with different inputs. [[1]](diffhunk://#diff-78df84c4c6e9e0905dd6dad83698db0e0d10113bee788bf552dfc2a1976bc8c2L312-R320) [[2]](diffhunk://#diff-78df84c4c6e9e0905dd6dad83698db0e0d10113bee788bf552dfc2a1976bc8c2R333-R390)

* [`src/promptflow-devkit/promptflow/_sdk/_utils/signature_utils.py`](diffhunk://#diff-a0108c48094370f09e9ce8192c56ed51fc4385d67b8b3ce0c27176b4bbc5a51bL155-R159): Modified the `update_signatures` function to allow for partial signatures in the YAML file.

* [`src/promptflow/tests/test_configs/eager_flows/basic_with_yaml_default/flow.flex.yaml`](diffhunk://#diff-4b52435378e0907fe710472385cff829af4c48dcf93e87c724fc87df9b6b52bfR1-R12): Added a new flow configuration with default values for `func_input` and `obj_input`.

* [`src/promptflow/tests/test_configs/eager_flows/basic_with_yaml_default/init.json`](diffhunk://#diff-8d46c736340e0198b67a746f330569aaa2fcbed48bf2bebc78619009d7ea5da9R1): Added a new initialization file for the `basic_with_yaml_default` flow.

* [`src/promptflow/tests/test_configs/eager_flows/basic_with_yaml_default/inputs.jsonl`](diffhunk://#diff-643e32a066ac8d210f374501f17aa5f3d5750ed1383d33e5c9b0d528f2a4ce08R1-R4): Added a new input file for the `basic_with_yaml_default` flow.

* [`src/promptflow/tests/test_configs/eager_flows/basic_with_yaml_default/run.yaml`](diffhunk://#diff-4df9f6e5317553efab07482c65935884ce5de6181cebf04c50c7cf624cde645cR1-R5): Added a new run configuration for the `basic_with_yaml_default` flow.

# All Promptflow Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes].**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [ ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: src/promptflow-core/promptflow/_utils/flow_utils.py [hunk 1]
```diff
@@ -152,16 +152,29 @@ def load_flow_dag(flow_path: Path):
     return flow_path, flow_dag
 
 
-def dump_flow_dag(flow_dag: dict, flow_path: Path):
-    """Dump flow dag to given flow path."""
-    flow_dir, flow_filename = resolve_flow_path(flow_path, check_flow_exist=False)
+def dump_flow_dag_to_existing_path(flow_dag: dict, flow_path: Path):
```
Comment (wangchao1230) [inline src/promptflow-core/promptflow/_utils/flow_utils.py:155]: raname: flow_dag -> flow yaml

File: src/promptflow-core/promptflow/contracts/flow.py [hunk 2]
```diff
@@ -459,6 +459,39 @@ def deserialize(data: dict):
...
+            elif data_type in MODEL_CONFIG_NAME_2_CLASS:
+                return data_type
+            return ValueType(data_type)
+
+        return FlowInitDefinition(
+            _get_type(data["type"]),
+            data.get("default", None),
+            data.get("description", ""),
+            data.get("enum", []),
+            data.get("is_chat_input", False),
```
Comment (wangchao1230) [inline src/promptflow-core/promptflow/contracts/flow.py:490]: do we have is_chat_input in Init Defintion?

File: src/promptflow-devkit/promptflow/_sdk/_orchestrator/utils.py [hunk 3]
```diff
@@ -195,11 +200,32 @@ def overwrite_flow(flow_dag: dict, params_overrides: dict):
 def remove_additional_includes(flow_path: Path):
     flow_path, flow_dag = load_flow_dag(flow_path=flow_path)
     flow_dag.pop("additional_includes", None)
-    dump_flow_dag(flow_dag, flow_path)
+    dump_flow_dag_to_existing_path(flow_dag, flow_path)
+
+
+def override_flow_dag(
```
Comment (wangchao1230) [inline src/promptflow-devkit/promptflow/_sdk/_orchestrator/utils.py:206]: override_flow_dag -> override_flow_yaml?

File: src/promptflow-core/promptflow/contracts/flow.py (hunk 2 above)
Comment (D-W-) [reply src/promptflow-core/promptflow/contracts/flow.py]: Extracted a base class

**Review by elliotzh — APPROVED** [review]

**Review by lumoslnt — APPROVED** [review]

**Review by lumoslnt — APPROVED** [review]

**Review by elliotzh — DISMISSED** [review]

**Review by crazygao — APPROVED** [review]

**Review by wangchao1230 — APPROVED** [review]

**Pull request merged**